xen/arch/x86/slaunch.c

```diff
@@ -0,0 +1,316 @@
+#include <xen/types.h>
+#include <asm/e820.h>
+#include <asm/intel_txt.h>
+#include <asm/page.h>
+#include <asm/processor.h>
+#include <asm/slaunch.h>
+#include <asm/tpm.h>
+#include <xen/init.h>
+#include <xen/mm.h>
+#include <xen/multiboot.h>
+
+/* SLB is 64k, 64k-aligned */
+#define SKINIT_SLB_SIZE  0x10000
+#define SKINIT_SLB_ALIGN 0x10000
+
+bool __initdata slaunch_active;
+uint32_t __initdata slaunch_slrt;
+
+static void __maybe_unused compile_time_checks(void)
+{
+    BUILD_BUG_ON(sizeof(slaunch_active) != 1);
+}
+
+int __init map_l2(unsigned long paddr, unsigned long size)
+{
+    unsigned long aligned_paddr = paddr & ~((1ULL << L2_PAGETABLE_SHIFT) - 1);
+    unsigned long pages = ((paddr + size) - aligned_paddr);
+    pages = ROUNDUP(pages, 1ULL << L2_PAGETABLE_SHIFT) >> PAGE_SHIFT;
+
+    if ( (aligned_paddr + pages * PAGE_SIZE) <= PREBUILT_MAP_LIMIT )
+        return 0;
+
+    if ( aligned_paddr < PREBUILT_MAP_LIMIT )
+    {
+        pages -= (PREBUILT_MAP_LIMIT - aligned_paddr) >> PAGE_SHIFT;
+        aligned_paddr = PREBUILT_MAP_LIMIT;
+    }
+
+    return map_pages_to_xen((unsigned long)__va(aligned_paddr),
+                            maddr_to_mfn(aligned_paddr),
+                            pages, PAGE_HYPERVISOR);
+}
+
+static uint32_t get_slb_start(void)
+{
+    /* The runtime computation relies on size being a power of 2 and equal to
+     * alignment. Make sure these assumptions hold. */
+    BUILD_BUG_ON(SKINIT_SLB_SIZE != SKINIT_SLB_ALIGN);
+    BUILD_BUG_ON(SKINIT_SLB_SIZE == 0);
+    BUILD_BUG_ON((SKINIT_SLB_SIZE & (SKINIT_SLB_SIZE - 1)) != 0);
+
+    /* Rounding any address within SLB down to alignment gives SLB base and
+     * SLRT is inside SLB on AMD. */
+    return slaunch_slrt & (SKINIT_SLB_SIZE - 1);
```

Comment:
Another sneaky error that I missed:
```suggestion
    return slaunch_slrt & ~(SKINIT_SLB_SIZE - 1);
```

```
(XEN) SLAUNCH: reserving SLB (0xd4d4 - 0x1d4d4)
```

Reply:
https://github.com/TrenchBoot/xen/compare/eb394884c439fca26cc5814cff59adfa10b21074..8ae81e157bfd4fa36c6bdb1dbf104d47edd5ce29